Add support for custom server RCON commands

maxijabase opened the pull request:
Duplicate of #376 from a feature branch.

maxijabase commented:
Currently the only macro I'm supporting is `{plugin}`, which will get replaced by the plugin that was compiled.

maxijabase commented:
- Added support for custom commands to be sent to the server via RCON. Placeholder macro available: `${plugin}`
- Fixed some cases of the code editor losing focus when compiling, and other related bugs
- Added progress, confirmation and error notifications when uploading files via (S)FTP or sending RCON commands.
- Fixed `compiler.path` setting not being used to locate the compiler. `spCompPath` should be properly deprecated now.

Source code-wise
- Changed `ftp-deploy` package for `ssh2-sftp-client` for SFTP and `basic-ftp` for FTP. They are a little more verbose (don't accept `grub` filters by default) but they had types for nicer development - we ought to see if the tradeoff is worth it.
- Polished config access (user settings) to simplify code.

TODO
- Make sure RCON command execution happens only if the Upload command promise has been fulfilled
- Fix notifications overlapping and disappearing for some confirmations
- Add nicer and clearer separations to the extension configuration
- Test ftp refactors more thoroughly
- Replace all "Open Settings" occurrences in notifications with more precise setting IDs for better UX instead of `@ext:sarrus.sourcepawn-vscode`
- Keep adding comments and checking the rest of the code for improvements

maxijabase commented:
- Fixed `upload` or `run commands` commands executing even if the compilation failed.
- Fixed `run commands` command running when `afterCompile` was set only if `uploadAfterSuccessfulCompile` was set to `true`.
- Removed progress notification in `run commands` command to avoid weird notifications behavior.

Probably a good idea to review the settings' names of the `upload` and `run commands` commands execution order/preferences.

Sarrus1 commented:
> Probably a good idea to review the settings' names of the `upload` and `run commands` commands execution order/preferences.

Could be `runServerCommandsOnSuccessfulUpload` ?

maxijabase commented:
- `changeSMApi` command: added check for empty APIs array, or any API with `name === ""` before rendering list.
- Added an `editConfig()` in the code similar to `getConfig()`, and used the setting ID to lead the user to the specific setting triggering the notification easier.
- Added `outputDirectoryPath` as a property of `AvailableAPIs` (#379)

maxijabase commented:
What's the python Scraper for?

Sarrus1 commented:
> What's the python Scraper for?

It used to be an attempt at parsing the CS:GO Cvars to be able to autocomplete them in cfg files

Sarrus1 commented on editors/code/package.json:
```diff
@@ -210,14 +210,8 @@
     ],
     "configuration": {
       "type": "object",
-      "title": "SourcePawn configuration",
```
nit: I prefer that this stays as `SourcePawn`

maxijabase replied:
Yes, I just followed this
https://code.visualstudio.com/api/references/contribution-points#:~:text=For%20both%20the,Extension%20Configuration%20Settings%22

Sarrus1 replied:
_"For both the title and displayName fields, words like "Extension", "Configuration", and "Settings" are redundant."_, so it should be `SourcePawn` right?

maxijabase replied:
Yes, exactly.

Sarrus1 replied:
Oh no, I read the diff backwards, I thought you added the `configuration` part... Soery about that haha

Sarrus1 commented on editors/code/package.json:
```diff
@@ -243,16 +231,24 @@
             "afterUpload"
           ],
-          "description": "Should the plugins list on the server be refreshed with `sm plugins refresh` after a successful action in VSCode.",
+          "description": "Should the specified commands be executed on the server after a successful action in VSCode (default: sm plugins refresh).",
```
I don't think it's necessary to mention the default of another setting here. I understand that they are related, but this might confuse some people

maxijabase replied:
Oh I think I made a mistake, if I'm not wrong this clarification was supposed to go in the description of the new setting that actually contains the commands, but you're right, defaults are already there, so I should remove it altogether.

Sarrus1 commented on editors/code/package.json:
```diff
@@ -243,16 +231,24 @@
+            "sm plugins refresh"
+          ],
+          "scope": "resource",
+          "description": "A list of commands that will be sent to the server after a successful action in VSCode."
```
```suggestion
          "description": "A list of commands that will be sent to the server after a successful VSCode command or on `runServerCommands`."
```

Sarrus1 commented on editors/code/package.json:
```diff
@@ -311,27 +284,20 @@
+            "username": "",
+            "password": "",
+            "sftp": false,
+            "remoteRoot": "/tf/addons/sourcemod",
```
Is this default intended?

maxijabase replied:
Yes, it actually is intended, opinionated assumption. Do you want another value there?
Also, I honestly don't know why I removed the `localRoot` variable and hardcoded it to the workspace root. Should I bring this back?

Sarrus1 replied:
In that case, `"/tf/addons/sourcemod"` is fine. I think workspace root is fine. The main point is to make it very hard for someone to accidentally delete their remote directory with this command.

Have you tested if there is some overwriting going on by default?

maxijabase replied:
Hm no. will do and report back.

Sarrus1 replied:
Thanks!

maxijabase replied:
If you were worried about files getting deleted from the remote location, if they were NOT present in the source, it's not happening in either of the FTP/SFTP libraries, which means it's not a directory synchronization. The only thing that happens is that existing files in destination are overwritten.

Sarrus1 commented on editors/code/package.json:
```diff
@@ -345,7 +311,7 @@
             "password": ""
           },
-          "description": "Options for the Source server to refresh the plugins on.",
+          "description": "Source server details to execute the commands on.",
```
Love the edit, much clearer!

Sarrus1 commented on editors/code/src/Commands/changeSMApi.ts:
```diff
@@ -1,80 +1,114 @@
+
+  if (!validApis) {
+    window.showInformationMessage(
+      "API list is empty or contains invalid entries! They must have a 'name' property.",
```
```suggestion
      "API list is empty or contains invalid entries. They must have a 'name' property.",
```

Sarrus1 commented on editors/code/src/Commands/changeSMApi.ts:
```diff
@@ -1,80 +1,114 @@
+
+  if (repeatedNames) {
+    window.showErrorMessage(
+      "API list has elements with duplicate names!",
```
```suggestion
      "API list has elements with duplicate names.",
```

Sarrus1 commented on editors/code/src/Commands/compileSM.ts:
```diff
@@ -23,90 +23,65 @@ let output: OutputChannel;
-  if (extname(fileToCompilePath) !== ".sp") {
-    window.showErrorMessage("Not a .sp file, aborting");
+  if (!isSPFile(fileToCompilePath)) {
+    window.showErrorMessage("The selected file is not a SourcePawn script.");
```
```suggestion
    window.showErrorMessage("The current file is not a SourcePawn script.");
```

Running the command with Ctrl+shift+D does not really "select" the file.

Sarrus1 commented on editors/code/src/Commands/compileSM.ts:
```diff
@@ -1,18 +1,18 @@
 import {
-  workspace as Workspace,
+  workspace,
```
Please keep it with a capital or import the whole `vscode` module. The reason is that `workspace` can be a variable in the file and this can cause name collisions.

maxijabase replied:
Understood!

Sarrus1 commented on editors/code/src/Commands/createCHANGELOG.ts:
```diff
@@ -2,31 +2,34 @@ import { workspace as Workspace, window, extensions } from "vscode";
 
-export function run(rootpath?: string) {
-  // get workspace folder
+export function run(rootpath?: string): void {
```
The functions return non zero values when they error out. This is used in automated tests.

Sarrus1 commented on editors/code/src/Commands/createGITIGNORE.ts:
```diff
@@ -2,31 +2,32 @@ import { workspace as Workspace, window, extensions } from "vscode";
 
-export function run(rootpath?: string) {
-  // get workspace folder
+export function run(rootpath?: string): void {
```
Same thing with the returns here.

Sarrus1 commented on editors/code/src/Commands/createGitHubActions.ts:
```diff
@@ -2,42 +2,39 @@ import { workspace as Workspace, window, extensions } from "vscode";
 
-export function run(rootpath?: string) {
-  // get workspace folder
+export function run(rootpath?: string): void {
```
Same thing with the returns here.

Sarrus1 commented on editors/code/src/Commands/createLICENSE.ts:
```diff
@@ -2,24 +2,24 @@ import { workspace as Workspace, window, extensions } from "vscode";
 
-export function run(rootpath?: string) {
-  // get workspace folder
+export function run(rootpath?: string): void {
```
Same thing with the returns here.

Sarrus1 commented on editors/code/src/Commands/createREADME.ts:
```diff
@@ -1,46 +1,39 @@
-        "You didn't specify a GitHub username.",
-        "Open Settings"
-      )
+export function run(rootpath?: string): void {
```
Same thing with the returns here.

Sarrus1 commented on editors/code/src/Commands/createScript.ts:
```diff
@@ -14,54 +13,55 @@ import {
-    "AuthorName"
-  );
-  if (!AuthorName) {
+export function run(rootpath?: string): void {
```
Same thing with the returns here.

Sarrus1 commented on editors/code/src/Commands/createTask.ts:
```diff
@@ -1,96 +1,107 @@
+import { editConfig, getConfig, Section } from "../configUtils";
 
-export function run(rootpath?: string) {
+export function run(rootpath?: string): void {
```
Same thing with the returns here.

Sarrus1 commented on editors/code/src/Commands/syntaxTree.ts:
```diff
@@ -41,16 +41,16 @@ export function syntaxTreeCommand(ctx: CtxInit): Cmd {
       const params: SyntaxTreeParams = {};
       const doc = vscode.window.activeTextEditor?.document;
-      if (doc === undefined) {
+      if (!doc) {
```
Please revert this. Using a negative expression on types can lead to weird JS behaviours which have cause problems in the past.

maxijabase replied:
Got it!

Sarrus1 commented on editors/code/src/Commands/syntaxTree.ts:
```diff
@@ -41,16 +41,16 @@ export function syntaxTreeCommand(ctx: CtxInit): Cmd {
       params.textDocument =
         ctx?.client.code2ProtocolConverter.asTextDocumentIdentifier(doc);
-      if (params.textDocument === undefined) {
+      if (!params.textDocument) {
```
Same here.

Sarrus1 commented on editors/code/src/Commands/syntaxTree.ts:
```diff
@@ -41,16 +41,16 @@ export function syntaxTreeCommand(ctx: CtxInit): Cmd {
       }
       const text = await ctx?.client.sendRequest(syntaxTree, params);
-      if (text === undefined) {
+      if (!text) {
```
Same here.

Sarrus1 commented on editors/code/src/spIndex.ts:
```diff
@@ -122,8 +123,19 @@ export async function activate(context: vscode.ExtensionContext) {
+    vscode.window.onDidChangeActiveTextEditor((editor) => {
+      // We check for a valid path in the editor's filename,
+      // which would indicate we're not on an output console
+      if (!path.isAbsolute(editor.document.fileName)) {
```
I think a better way to check for this is through the scheme of the document's uri?

maxijabase replied:
Hm can you illustrate a bit? What exactly would you check? Parse the filename as URI and extract stuff or something?

Sarrus1 replied:
```suggestion
      if (editor.document.uri.scheme === "output") {
```
More generally, the scheme of a URI is a specification that allows to assign reserved characters to the rest of the URI.
In our case, the relevant part is that a document's URI scheme can be a `file` or `output` (and others), `output` being for the output console.

maxijabase replied:
Cool!

Sarrus1 commented on editors/code/src/spIndex.ts:
```diff
@@ -152,7 +171,7 @@ export function getCtxFromUri(uri: URI): Ctx | undefined {
   // If we have nested workspace folders we only start a server on the outer most workspace folder.
   folder = getOuterMostWorkspaceFolder(folder);
-  console.log(serverContexts);
+  // console.log(serverContexts);
```
Can be removed :)

Sarrus1 commented on editors/code/src/spUtils.ts:
```diff
@@ -1,11 +1,33 @@
-export function isSPFile(fileName: string) {
-  return /(?:\.sp|\.inc)\s*^/.test(fileName);
+export function isSPFile(filePath: string): boolean {
+  return /\.(sp|inc)$/i.test(filePath);
```
I don't think the compiler is case insensitive?

maxijabase replied:
Hm I think I changed that regex because it wasn't working, I tested it in regex101.com but it wasn't matching with anything that function was receiving, but I could be wrong.

Sarrus1 replied:
The trailing `^` was 100% wrong, that's for sure, but you can remove the `i` in the new one.

Sarrus1 replied:
```suggestion
  return /\.(sp|inc)$/.test(filePath);
```

Sarrus1 reviewed (commented):
Thanks again for this PR, it's massive! A lot of this was overdue... There are many comments given the scope of the PR.
